File: codes/Haskell/p021.hs
{- 
 - Solution to Project Euler problem 21
 - Copyright (c) Sankha Saha. All rights reserved.
 - 
 -  
 -   
 -}


{- 
 - We find the sum of proper divisors of a number by brute force,
 - and apply the definition of an amicable number straightforwardly.
 -}

main = putStrLn (show ans)
ans = sum [n | n <- [1..10^4], amicable n]

amicable :: Int -> Bool
amicable n = let m = divisorSum n in (m /= n) && (divisorSum m) == n

divisorSum :: Int -> Int
divisorSum n = sum [k | k <- [1..n-1], (mod n k) == 0]

File: codes/mathematica/p021.m
(* 
 * Solution to Project Euler problem 21
 * Copyright (c) Sankha Saha. All rights reserved.
 * 
 *  
 *   
 *)


(* 
 * We modify a built-in function to find the sum of proper divisors of a number.
 * Then we apply the definition of an amicable number straightforwardly.
 *)
DivisorSum[n_] := DivisorSigma[1, n] - n
AmicableQ[n_] := DivisorSum[n] != n && DivisorSum[DivisorSum[n]] == n
Total[Select[Range[9999], AmicableQ]]
